Framework/Core/src/Dispatcher.cxx

```diff
@@ -0,0 +1,157 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file Dispatcher.cxx
+/// \brief Implementation of Dispatcher for O2 Data Sampling
+///
+/// \author Piotr Konopka, piotr.jan.konopka@cern.ch
+
+#include "Framework/Dispatcher.h"
+#include "Framework/RawDeviceService.h"
+#include "Framework/DataSamplingPolicy.h"
+#include "Framework/DataProcessingHeader.h"
+
+#include <Configuration/ConfigurationInterface.h>
+#include <Configuration/ConfigurationFactory.h>
+#include <fairmq/FairMQDevice.h>
+#include <fairmq/FairMQLogger.h>
+
+using namespace o2::configuration;
+
+namespace o2
+{
+namespace framework
+{
+
+Dispatcher::Dispatcher(std::string name, const std::string reconfigurationSource)
+  : mName(name), mReconfigurationSource(reconfigurationSource)
+{
+}
+
+Dispatcher::~Dispatcher()
+{
+}
+
+void Dispatcher::init(InitContext& ctx)
+{
+  LOG(DEBUG) << "Reading Data Sampling Policies...";
+
+  std::unique_ptr<ConfigurationInterface> cfg = ConfigurationFactory::getConfiguration(mReconfigurationSource);
+  auto policiesTree = cfg->getRecursive("dataSamplingPolicies");
+  mPolicies.clear();
+
+  for (auto&& policyConfig : policiesTree) {
+    mPolicies.emplace_back(std::make_shared<DataSamplingPolicy>(policyConfig.second));
+  }
+}
+
+void Dispatcher::run(ProcessingContext& ctx)
+{
+  for (const auto& input : ctx.inputs()) {
+    if (input.header != nullptr && input.spec != nullptr) {
+
+      for (auto& policy : mPolicies) {
+        // todo: consider getting the outputSpec in match to improve performance
+        // todo: consider matching (and deciding) in completion policy to save some time
+        if (policy->match(*input.spec) && policy->decide(input)) {
+
+          if (!policy->getFairMQOutputChannel().empty()) {
+            sendFairMQ(ctx.services().get<RawDeviceService>().device(), input, policy->getFairMQOutputChannelName());
+          } else {
+            send(ctx.outputs(), input, policy->prepareOutput(*input.spec));
+          }
+        }
+      }
+    }
+  }
+}
+
+void Dispatcher::send(DataAllocator& dataAllocator, const DataRef& inputData, const Output& output) const
+{
+  //todo: support other serialization methods
+  const auto* inputHeader = header::get<header::DataHeader*>(inputData.header);
+  if (inputHeader->payloadSerializationMethod == header::gSerializationMethodInvalid) {
+    LOG(WARNING) << "DataSampling::dispatcherCallback: input of origin'" << inputHeader->dataOrigin.str
+                 << "', description '" << inputHeader->dataDescription.str
+                 << "' has gSerializationMethodInvalid.";
+  } else if (inputHeader->payloadSerializationMethod == header::gSerializationMethodROOT) {
+    dataAllocator.adopt(output, DataRefUtils::as<TObject>(inputData).release());
+  } else { // POD
+    // todo: do it non-copy, when API is available
+    auto outputMessage = dataAllocator.newChunk(output, inputHeader->payloadSize);
```

Comment:
@ktf @matthiasrichter Is there now any way to adopt input data safely, without a copy?